File: solutions/ruby/bob/1/bob.rb
module Bob
  def self.hey(remark) 
    remark.strip! 
    question = remark.match?(/\A[A-Za-z0-9!\.\?\-\+\*\^\\\|\$\(\)\[\]\{\},: ]+\?\Z/)
    yell = remark.match?(/(?!^[\d ,]+$)\A[A-Z \%\\^\*\@\\\#\$\(0-9!,]+\Z/)
    yell_question = remark.match?(/\A[A-Z\\.\\?\\-\\+\\*\\^\\|\\$\\(\\)\\[\\]\\{\\' ]+\?\Z/)
    silence = remark.match?(/^\s*$/)
    case 
    when question then "Sure."
    when yell then "Whoa, chill out!"
    when yell_question then "Calm down, I know what I'm doing!"
    when silence then "Fine. Be that way!" 
    else "Whatever." 
    end 
  end
end 

File: solutions/cpp/bob/1/bob.cpp
#include "bob.h"
#include<algorithm> 

using namespace std; 

namespace bob {
    bool isQuestion(const string& msg) {
        return msg.back() == '?';
    }

    bool lettersAreCaps(const string& msg) {
        int capsCount{0};
        for (const char c : msg) {
            if (isalpha(c)) {
                if (!isupper(c)) 
                    return false;
                else 
                    capsCount++;
            } else {
                continue; 
            }
        }
        
        return (capsCount > 0);
    }

    bool isSilence(const string& msg) {
        if (msg.empty()) return true;
        for (const char c : msg) {
            if (!isspace(c)) return false;
        }
        return true; 
    }

    string hey(string msg) {
        msg.erase(msg.find_last_not_of(" \t\n\r\f\v") + 1);
        string answer{"Whatever."};
        
        if (isSilence(msg)) {
            return "Fine. Be that way!";
        }
        
        if (lettersAreCaps(msg)) {
            if (isQuestion(msg))
                return "Calm down, I know what I'm doing!";
            else 
                return "Whoa, chill out!";
        }
        
        if (isQuestion(msg)) {
            return "Sure.";
        }
        
        return answer; 
    }

}  // namespace bob
